packages/react-native-reanimated/Common/cpp/reanimated/NativeModules/NativeReanimatedModule.cpp

```diff
@@ -704,72 +701,77 @@ void NativeReanimatedModule::performOperations() {
-
-  shadowTreeRegistry.visit(surfaceId_, [&](ShadowTree const &shadowTree) {
-    shadowTree.commit(
-        [&](RootShadowNode const &oldRootShadowNode)
-            -> RootShadowNode::Unshared {
-          PropsMap propsMap;
-          for (auto &[shadowNode, props] : copiedOperationsQueue) {
-            auto family = &shadowNode->getFamily();
-            react_native_assert(family->getSurfaceId() == surfaceId_);
-            propsMap[family].emplace_back(rt, std::move(*props));
-
-            if (propsRegistry_->shouldReanimatedSkipCommit()) {
-              return nullptr;
+  
+    
+    
+        bool hasLayoutUpdates = false;
+        
+        for (const auto &[shadowNode, props] : copiedOperationsQueue) {
+            if (isThereAnyLayoutProp(rt, props->asObject(rt))) {
+                hasLayoutUpdates = true;
+                break;
             }
-          }
-
-          auto rootNode =
-              cloneShadowTreeWithNewProps(oldRootShadowNode, propsMap);
-
-          // Mark the commit as Reanimated commit so that we can distinguish it
-          // in ReanimatedCommitHook.
-
-          auto reaShadowNode =
-              std::reinterpret_pointer_cast<ReanimatedCommitShadowNode>(
-                  rootNode);
-          reaShadowNode->setReanimatedCommitTrait();
-
-          return rootNode;
-        },
-        {/* .enableStateReconciliation = */
-         false,
-         /* .mountSynchronously = */ true,
-         /* .shouldYield = */ [this]() {
-           return propsRegistry_->shouldReanimatedSkipCommit();
-         }});
-  });
+        }
+        
+        if (!hasLayoutUpdates) {
+            // If there's no layout props to be updated, we can apply the updates
+            // directly onto the components and skip the commit.
+            for (const auto &[shadowNode, props] : copiedOperationsQueue) {
+                    Tag tag = shadowNode->getTag();
+                    synchronouslyUpdateUIPropsFunction_(rt, tag, props->asObject(rt));
+                    
+            }
+            return;
+        }
+        
+        if (propsRegistry_->shouldReanimatedSkipCommit()) {
+            // It may happen that `performOperations` is called on the UI thread
+            // while React Native tries to commit a new tree on the JS thread.
+            // In this case, we should skip the commit here and let React Native do
+            // it. The commit will include the current values from PropsRegistry which
+            // will be applied in ReanimatedCommitHook.
+            return;
+        }
+        
+        
+        react_native_assert(uiManager_ != nullptr);
+        const auto &shadowTreeRegistry = uiManager_->getShadowTreeRegistry();
+        
+    for (auto &[shadowNode, props] : copiedOperationsQueue) {
+        SurfaceId surfaceId_ = shadowNode->getSurfaceId();
+        shadowTreeRegistry.visit(surfaceId_, [&](ShadowTree const &shadowTree) {
```

Comment:
I'm not sure if this is an optimal approach here in terms of execution time – in particular, we don't know how expensive is a call to `shadowTreeRegistry.visit`. In the vast majority of apps, there will be only one surface so we should heavily optimize for that use case. If there are more surfaces, maybe we should group the updates by surface ID so we visit each surface exactly once?

Reply:
Grouping by surfaceId was my initial thought! I will change it accordingly